state.go

```diff
@@ -557,6 +560,114 @@ func (s *State) Channel(channelID string) (*Channel, error) {
 	return nil, ErrStateNotFound
 }
 
+// ThreadMemberUpdate updates the ThreadMember object for the user,
+// in a specific thread.
+func (s *State) ThreadMemberUpdate(t *ThreadMember) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	channel, ok := s.channelMap[t.ID]
+	if ok {
+		s.Lock()
+		defer s.Unlock()
+
+		channel.Member = t
+		return nil
+	}
+
+	return ErrStateNotFound
+}
+
+// ThreadMembersUpdate updates the member object of the thread.
+// If the user was removed,  the ThreadMember object in the state
+// will be set to nil.
+func (s *State) ThreadMembersUpdate(t *ThreadMembersUpdate) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	thread, ok := s.channelMap[t.ID]
+	if ok {
+		for _, memberID := range t.RemovedMembersIDs {
+			if memberID == s.User.ID {
+				thread.Member = nil
+				break
+			}
+		}
+
+		for _, member := range t.AddedMembers {
+			if member.UserID == s.User.ID {
+				thread.Member = member
+				break
+			}
+		}
+	}
+
+	return ErrStateNotFound
+}
+
+// ThreadListSync syncs new threads and threads member that
+// the bot gained access to.
+// https://discord.com/developers/docs/topics/gateway#thread-list-sync
+func (s *State) ThreadListSync(t *ThreadListSync) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	g, err := s.Guild(t.GuildID)
+	if err != nil {
+		return err
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	var index int
+
+OUTER:
+	for _, c := range g.Channels {
+		if !c.IsThread() || c.ThreadMetadata.Archived {
+			g.Channels[index] = c
+			index++
+			continue OUTER
+		}
+
+		if len(t.ChannelIDs) == 0 {
+			delete(s.channelMap, c.ID)
+			continue OUTER
+		}
+
+		for _, parentChannelID := range t.ChannelIDs {
+			if c.ParentID == parentChannelID {
+				delete(s.channelMap, c.ID)
+				continue OUTER
+			}
+		}
+
+		g.Channels[index] = c
+		index++
+	}
```

Comment:
A labeled loop with three different (and similar) `continue` points seems really difficult to reason about to me. I would suggest inverting the loop—first loop over the thread sync event's channels, and then do the guild channels in the inner loop, since you'll then be starting the guild channels loop anew each time and won't need to worry about mutating in place.